test-tools/src/main/java/org/terracotta/utilities/test/logging/ConsoleAppenderCapture.java

```diff
@@ -0,0 +1,235 @@
+   *     assertThat(logCapture, Matcher...);
+   *   }
+   * }</pre>
+   *
+   * @return a new {@code ConsoleAppenderCapture} instance
+   * @throws IllegalStateException if the current logging configuration does not append to {@link ConsoleAppender}
+   * @see ConsoleAppenderCapture#ConsoleAppenderCapture(String)  ConsoleAppenderCapture
+   */
+  public static ConsoleAppenderCapture capture() {
+    return ConsoleAppenderCapture.capture(null);
+  }
+
+  /**
+   * Creates a {@code ConsoleAppenderCapture} instance which captures events sent to each {@code Logger}
+   * in the current Logback configuration having a {@link ConsoleAppender}.  Only events logged by threads
+   * for which the {@link MDC} contains a key generated by this constructor are captured.  If multiple
+   * {@code ConsoleAppender} instances are used (uncommon), the captures from all instances having the
+   * same target are aggregated.
+   * <p>
+   * If the current logging configuration does not append to {@code ConsoleAppender}, no capture is
+   * performed and {@link #getLogs()} will return an empty map.
+   * <p>
+   * If a {@code Logger}, through inheritance, appends to multiple instances of {@code ConsoleAppender},
+   * then multiple event captures may be observed.
+   * <p>
+   * The {@code ConsoleAppenderCapture} instance <b>must</b> be closed when no longer needed -- use
+   * try-with-resources.
+   *
+   * @param testId used as the value for the generated {@code MDC} key; if {@code null}, a value of
+   *               {@code present} is used
+   */
+  public ConsoleAppenderCapture(String testId) {
+    this.testId = (testId == null ? "present" : testId);
+
+    MDC.put(mdcKey, this.testId);
+    Runnable removeAppenders = () -> {
+    };
+    LoggerContext context = (LoggerContext)LoggerFactory.getILoggerFactory();
+
+    /*
+     * Scan for all loggers using the ConsoleAppender and add an appender to capture
+     * the output for the designated ConsoleAppender target.
+     */
+    for (ch.qos.logback.classic.Logger logger : context.getLoggerList()) {
+      Iterator<Appender<ILoggingEvent>> appenderIterator = logger.iteratorForAppenders();
+      while (appenderIterator.hasNext()) {
+        Appender<ILoggingEvent> iLoggingEventAppender = appenderIterator.next();
+        if (iLoggingEventAppender instanceof ConsoleAppender) {
+          String target = ((ConsoleAppender<?>)iLoggingEventAppender).getTarget();
+          ListAppender<ILoggingEvent> appender = appenderMap.computeIfAbsent(target, t -> {
+            ListAppender<ILoggingEvent> listAppender = new ListAppender<>();
+            listAppender.setContext(context);
+
+            listAppender.addFilter(new AbstractMatcherFilter<ILoggingEvent>() {
+              @Override
+              public FilterReply decide(ILoggingEvent iLoggingEvent) {
+                if (iLoggingEvent.getMDCPropertyMap().getOrDefault(mdcKey, "").equals(ConsoleAppenderCapture.this.testId)) {
```

Comment:
if testId is null (== `present`), should be accept everything ?

Reply:
See the response above -- the `testId` isn't really relevant to capture ... it's solely for identification.